testsupport/wait/member.go

```diff
@@ -989,6 +989,48 @@ func (a *MemberAwaitility) WaitForRoleBinding(t *testing.T, namespace *corev1.Na
 	return roleBinding, err
 }
 
+// WaitForObject waits until an Object with the given name exists in the given namespace
+func (a *MemberAwaitility) WaitForObject(t *testing.T, namespaceName string, name string, obj client.Object, objList client.ObjectList) error {
+	t.Logf("waiting for Object '%s' in namespace '%s'", name, namespaceName)
+	err := wait.Poll(a.RetryInterval, a.Timeout, func() (done bool, err error) {
+		if err := a.Client.Get(context.TODO(), types.NamespacedName{Namespace: namespaceName, Name: name}, obj); err != nil {
```

Comment:
reusing the same object in the loop is risky. We should always use a new instance of the object (or we need to know exactly what resources for which purpose we are getting, which is hard in such a generic function). 
If we don't do that, then the resulting object may contain fields/conditions/values that are not present in the corresponding resource anymore. 
In other words, the final content of the object doesn't have to reflect what is in the cluster.

Reply:
This is just a theoretical risk in this specific case, isn't it? This specific function just gets the object by this name and if found it returns it. It doesn't check anything else like labels or conditions which would populate the given obj with some found object but instead of returning it would try to populate it again in a loop. This function should work as is but I agree that it's error prone in case we want to change it (and introduce criteria params for example) in the future.
Creating a deep copy would be safer but I'm not sure if there is any easy way of creating a deep copy of a client.Object. We could try to marshal/unmarshal it but not sure if it's worth doing...
Alternatively we could drop it and just add a `WaitForClusterResourceBinding()` function instead but I still find this generic `WaitForObject()` useful.
WDYT?

Reply:
In your specific case, yes. But this is a generic function that could be used for anything and you cannot control how the caller uses the function. 
I found an old PR where we were fixing this problem https://github.com/codeready-toolchain/toolchain-e2e/pull/43 and I guess that there were a few more cases since then.
Anyway, why couldn't we use the already existing generic functions? https://github.com/codeready-toolchain/toolchain-e2e/blob/82dc979b1fafe3bf61792736cad45b71baead750/testsupport/wait/awaitility.go#L657-L845
Instead of calling:
```go
crb := &v1.ClusterRoleBinding{}
crbName := fmt.Sprintf("%s-%s", user.Space.Name, "test-feature")
err = memberAwait.WaitForObject(t, "", crbName, crb, &v1.ClusterRoleBindingList{})
require.NoError(t, err)
```
you could do
```go
crbName := fmt.Sprintf("%s-%s", user.Space.Name, "test-feature")
crb, err := wait.For(t, memberAwait.Awaitility, &v1.ClusterRoleBinding{}).WithNameThat(crbName)
require.NoError(t, err)
```
or create a new method that would accept also the namespace because the current one automatically takes the operator's namespace - something like that:
```go
crbName := fmt.Sprintf("%s-%s", user.Space.Name, "test-feature")
crb, err := wait.For(t, memberAwait.Awaitility, &v1.ClusterRoleBinding{}).WithNamespacedNameThat("", crbName)
require.NoError(t, err)
```

TBH, I would rather make this generic `Waiter` more powerful and useful than creating yet another function that does more or less the same, but in a slightly risky way and less powerful way. I understand that the `Waiter` may look complicated and doesn't provide the nice logging as the other wait functions, but take it as the start. We can make it better. Mainly the predicate abstraction could have been done in a more useful way.

Reply:
> In your specific case, yes. But this is a generic function that could be used for anything and you cannot control how the caller uses the function.
I found an old PR where we were fixing this problem https://github.com/codeready-toolchain/toolchain-e2e/pull/43 and I guess that there were a few more cases since then.

It's not actually about how this function can be used but more about how it's implemented. The PR/fix you provided is a bit different because, as I mentioned above, the function accepts criteria. So the function loads the object multiple times in the same struct until it satisfies the criteria. 

But the proposed `WaitForObject()` function does not accept any criteria. So it doesn't load the struct multiple times. As soon as the object is found it returns it as is. Previous attempts should not affect it. The only potential issue can be if you pass non-empty object to that function. But it's no different to using `client.Get(object)`.

The potential risk I was referring to is related to situation if we decide to add some criteria to the function later without taking into account the implementation details. In this case we would introduce a bug into the function.

But anyway. I was not aware of that `wait.For()` function and this is why I introduced `WaitForObject()`. So let me drop this and re-use the existing function: https://github.com/codeready-toolchain/toolchain-e2e/pull/1008/commits/df30805f96d33bf08e9d2eff8bad4ffc1d4334d0
I also added `wait.For().WithNameDeleted()` instead of `WaitForObjectToBeDeleted()`.

Reply:
you are actually right, the for loop would immediately stop when the resources is present, so there wouldn't be multiple Get calls that would get different versions of the resources. I had the mindset of multiple executions waiting from some criteria, sorry for confusion :+1: 

Anyway, thanks for reusing the `wait.For()` capability :+1: